### PR TITLE
Refactor how to handle the dataset and add support for CIFAR10

### PR DESCRIPTION
**Pull request by bowni**
Addresses issue #139, #161.

### CODE REVIEW

File: constants.py [hunk 1]
```diff
@@ -3,6 +3,9 @@
 Declaration of constants.
 """
 
+# Dataset
+DATASET = str
```
Comment (RomainGoussault) [inline constants.py:7]: constant should be constant and already initialized

File: dataset.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
+# -*- coding: utf-8 -*-
+"""
+The dataset object used in the multi-partner learning and contributivity measurement experiments.
+"""
+
+from sklearn.model_selection import train_test_split
+
+
+class Dataset:
```
Comment (Thomas-Galtier) [inline dataset.py:9]: I would rather call the class Dataset_and_model

File: datasets/dataset_cifar10.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+y_train = y_train.flatten()
+y_test = y_test.flatten()
+
+# Init dataset-specific variables
+input_shape = (32, 32, 3)
+num_classes = 10
+
+# Pre-process inputs
+x_train = preprocess_dataset_inputs(x_train)
+x_test = preprocess_dataset_inputs(x_test)
```
Comment (Thomas-Galtier) [inline datasets/dataset_cifar10.py:41]: I wonder why are the data are preprocessed here? It would be clearer to preprocess them  where the label are preprocessed

File: multi_partner_learning.py [hunk 4]
```diff
@@ -505,7 +500,7 @@ def log_collaborative_round_partner_result(self, partner, partner_index, validat
 
         epoch_nb_str = f"Epoch {str(self.epoch_index).zfill(2)}/{str(self.epoch_count - 1).zfill(2)}"
         mb_nb_str = f"Minibatch {str(self.minibatch_index).zfill(2)}/{str(self.minibatch_count - 1).zfill(2)}"
-        partner_id_str = f"Partner id #{partner.id} ({partner_index}/{self.partners_count})"
+        partner_id_str = f"Partner id #{partner.id} ({partner_index}/{self.partners_count - 1})"
```
Comment (natct10) [inline multi_partner_learning.py:503]: oh!

File: datasets/dataset_cifar10.py (hunk 3 above)
Comment (natct10) [reply datasets/dataset_cifar10.py:41]: Ah, I didn't see your comment

Comment (bowni) [reply datasets/dataset_cifar10.py:41]: Kind of a dilemma:
- I think the best would be that the user providing the dataset provides it all pre-processed. But thing is, for the labels it is not ideal, as we make quite some stuff with them during the data split. So the labels we have to put to binary vector after the data split
- So either we keep the input pre-processing in the datasets dedicated files, so that the user "delivers" clean inputs, or we put it after the data split but that entangles a bit more the dataset and the library logic

Comment (Thomas-Galtier) [reply datasets/dataset_cifar10.py:41]: ok we can let it like that then

File: datasets/dataset_mnist.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+# Data samples pre-processing method for labels
+def preprocess_dataset_labels(y):
+
+    y = keras.utils.to_categorical(y, num_classes)
+
+    return y
+
+
+# Load data
+(x_train, y_train), (x_test, y_test) = mnist.load_data()
```
Comment (natct10) [inline datasets/dataset_mnist.py:33]: cc @Fabien-GELUS

File: dataset.py (hunk 2 above)
Comment (bowni) [reply dataset.py:9]: Ok

**Comment by bowni** [comment]
@RomainGoussault can you do a final review and merge if OK?

File: README.md [hunk 6]
```diff
@@ -258,7 +258,7 @@ Example: `is_quick_demo: True`
 
 - [**Shapley values**](https://arxiv.org/pdf/1902.10275.pdf):  
 
-  These indicators seem to be very good candidates to measure the contributivity of each data providers, because they are usually used in game theory to fairly attributes the gain of a coalition game amongst its players, which is exactly want we are looking for here.
+  These indicators seem to be very good candidates to measure the contributivity of each data providers, because they are usually used in game theory to fairly attributes the gain of a coalition game amongst its players, which is exactly what we are looking for here.
```
Comment (natct10) [inline README.md:261]: :+1:

File: README.md [hunk 7]
```diff
@@ -100,18 +104,18 @@ Finally, with given scenarios and multi-partner learning approaches, we can addr
...
+   Under `scenario_params_list`, enter a list of sets of scenario(s). Each set starts with ` - dataset_name:` and must have only one `partners_count` value. The length of `amount_per_partners`, `corrupted_datasets` (and `samples_split_option` when the advanced definition is used) must match the `partner_counts` value. If for a given parameter multiple values are specified, e.g. like for `agregation_weighting` in the first scenario set of the above example, all possible combinations of parameters will be assembled as separate scenarios and run.
    
 3. Then execute `main.py -f config.yml`
 
-4. A `results.csv` file will be generated in a new folder for your experiment under `/experiments/<your_experiment>`. You can read this raw `results.csv` file or use the `analyse_results.ipynb` generated notebook to quickly generate figures.
+4. A `results.csv` file will be generated in a new folder for your experiment under `/experiments/<your_experiment>`. You can read this raw `results.csv` file or use the generated notebooks in `/notebooks`.
```
Comment (RomainGoussault) [inline README.md:118]: or use the notebooks in `/notebooks`.

The notebooks are not re-generated at every experiment

File: scenario.py [hunk 8]
```diff
@@ -16,26 +15,49 @@
...
 
 class Scenario:
     def __init__(self, params, experiment_path, scenario_id=1, n_repeat=1):
 
-        # Identify and get a dataset for running experiments
-        self.dataset_name = "MNIST"
-        (x_train, y_train), (x_test, y_test) = mnist.load_data()
-        self.nb_samples_used = len(x_train)
+        # Get and verify which dataset is configured
+        supported_datasets = ["mnist", "cifar10"]
```
Comment (RomainGoussault) [inline scenario.py:27]: `supported_dataset_names` to be consistent. 
Or `dataset `instead of `dataset_name`

File: dataset_and_model.py [hunk 9]
```diff
@@ -0,0 +1,52 @@
+# -*- coding: utf-8 -*-
+"""
+The dataset object used in the multi-partner learning and contributivity measurement experiments.
+"""
+
+from sklearn.model_selection import train_test_split
+
+
+class DatasetAndModel:
```
Comment (RomainGoussault) [inline dataset_and_model.py:9]: I would rename it to `Dataset `since it's mostly dataset related (there is only one function which relates to a model). And when you instantiate it, you name it `dataset_mnist `or `dataset_cifar`.

File: datasets/dataset_cifar10.py (hunk 3 above)
Comment (RomainGoussault) [inline datasets/dataset_cifar10.py:41]: I don't understand where is this used ?

File: README.md (hunk 7 above)
Comment (bowni) [reply README.md:118]: Ok

File: scenario.py (hunk 8 above)
Comment (bowni) [reply scenario.py:27]: Ok

File: dataset_and_model.py (hunk 9 above)
Comment (bowni) [reply dataset_and_model.py:9]: Ok

File: datasets/dataset_cifar10.py (hunk 3 above)
Comment (bowni) [reply datasets/dataset_cifar10.py:41]: Just here: inputs are pre-processed on the spot here.

Comment (RomainGoussault) [reply datasets/dataset_cifar10.py:41]: Ok I understand now !

**Comment by bowni** [comment]
@RomainGoussault for you again :-)
